### DISCUSSION TITLE
JS-SDK: Synchronize directus auth information between client and server (SSR)

### DISCUSSION THREAD

**Discussion by henk23** [directus/directus#14174] [Archive]
I'm writing a Next.js app and currently cannot wrap my head around how to handle directus authentication both on client side and during server side rendering (using getServerSideProps()).

Currently this is what I have come up with:

- In a server side /api/login endpoint,
  - call directus.auth.login()
  - save the response to the session,
  - return the response to the client,
  - on client side, call directus.storage.set() for the auth_token, auth_refresh_token, auth_expires, auth_expires_at
- In a server side /api/logout endpoint,
  - call directus.auth.logout()
  - destroy the session
  - on client side, call directus.storage.delete() for all the values.

Now, this would almost work, but: At any point on the server or on the client, directus will automatically refresh the token, and then I will have to somehow sync the new auth token to the respective other side.

Does anyone have a good solution for this scenario? Would be nice to find an example or directions in the docs.

**Comment by benhaynes** [comment]
@azrikahar Does our Next.js codebase include an example of auth?

https://directus.io/guides/get-started-building-a-next-website-with-directus/

**Reply by azrikahar** [reply]
The codebase does authenticate when fetching the data, but our example currently uses `getStaticProps` (static site generation), so it technically only needs to do the auth server side one time.

In order for `getServerSideProps` (server side rendering) to work more seamlessly in OP's case, I believe using access token via request header or cookie (some users have implemented their own custom endpoints for this) and use that token _server-side_ should have less moving parts than the current approach described in OP.

**Reply by henk23** [reply]
The example repo depends on email&password via env variables. Not really viable for a multi user scenario.

I came to the conclusion that my approach won't work, so I went with:

- SSR with getServerSideProps for all public, unauthenticated pages, making the items publicly available in directus.
- Client side data fetching with useEffect for authenticated users in the log-in area.

**Reply by sooqwaifr** [reply]
@azrikahar 
But how to set the access token in the header and later use it in getserverSideProps?.
Right now only the refresh token is stored in the cookie and you can access it in getServersideProps request.
Is there away to store the access token as well when the user login or the token is refresh?
I believe the js-sdk client should have two aditional features:

- Ability to subscribe to auth changes.  add onAuthStateChange, so you can trigger an api callback to set the access cookie in the header.
- Ability to set the access token manually. so you can get the token from the request server side and pass it the sdk.

directus.auth.onAuthStateChange()
directus.auth.setAuth(token)

**Reply by azrikahar** [reply]
> But how to set the access token in the header and later use it in getserverSideProps?.

Granted I'm not 100% sure here, but I believe the access token (if the cooke is set) should be sent in the request header if you are using axios withCredentials or fetch [credentials](https://developer.mozilla.org/en-US/docs/Web/API/fetch#credentials).

Then it should be possible to read the request header in getServerSideProps' [context](https://nextjs.org/docs/api-reference/data-fetching/get-server-side-props#context-parameter), if I'm not mistaken.

> Is there away to store the access token as well when the user login or the token is refresh?

There isn't an out-of-the-box way, but I recall some users have created a custom endpoint to set the access token header. Regarding including it by default, a related discussion could be #10103.

> I believe the js-sdk client should have two aditional features

These could be interesting features, but there may be non-trivial hurdles when it comes to hybrid renderings. That said, as the SDK has been moved to a standalone repo over here: https://github.com/directus/sdk, it could be worth making a formal suggestion in a discussion over there 👍
